feat: use `SoftFailure` in `Weeder2` at `pickRestrictableChoosePattern`

jaschdoc opened the pull request:
Related to https://github.com/flix/flix/issues/8540

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:
```diff
@@ -1389,16 +1389,27 @@ object Weeder2 {
+        case other =>
+          val e = UnsupportedRestrictedChoicePattern(isStar, other.loc)
+          val ident = Name.Ident(s"x${flix.genSym.freshId()}", other.loc.asSynthetic)
+          val patVar = WeededAst.RestrictableChoosePattern.Var(ident, other.loc)
```
Why are we using `Var` here instead of a proper error node?

magnus-madsen replied:
Yeah, we should definitely have a proper error node.

jaschdoc replied:
Tried adding an error node as a pattern but I can also try to do it as a var

jaschdoc commented:
Tried with the var in https://github.com/flix/flix/pull/8608

jaschdoc commented:
!!!rerun

magnus-madsen commented:
@jaschdoc Flixball is failing for unrelated reasons.

mlutze commented:
!!!rerun

magnus-madsen commented:
RFR?

jaschdoc commented:
Yes

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:
```diff
@@ -1389,16 +1389,24 @@ object Weeder2 {
+                case other =>
+                  val e = UnsupportedRestrictedChoicePattern(isStar, other.loc)
+                  val ident = Name.Ident(s"x${flix.genSym.freshId()}", other.loc.asSynthetic)
+                  Validation.toSoftFailure(WeededAst.RestrictableChoosePattern.Var(ident, other.loc.asSynthetic), e)
```
This should use the error construct right?

jaschdoc replied:
See the other PR i opened

magnus-madsen commented on main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:
```diff
@@ -1389,16 +1389,24 @@ object Weeder2 {
+            case other =>
+              val e = UnsupportedRestrictedChoicePattern(isStar, other.loc)
+              val ident = Name.Ident(s"x${flix.genSym.freshId()}", other.loc.asSynthetic)
+              Validation.toSoftFailure(List(WeededAst.RestrictableChoosePattern.Var(ident, other.loc.asSynthetic)), e)
```
ditto

magnus-madsen commented:
Wait. This is outdated right?

magnus-madsen commented:
Or not. I am so confused.

jaschdoc commented:
2 types of error nodes. Here I added one where Tag would be used and in the other PR I added error nodes instead of vars. I'm not sure which one is preferred or if we need both (hence two PRs) 😀

magnus-madsen commented:
> 2 types of error nodes. Here I added one where Tag would be used and in the other PR I added error nodes instead of vars. I'm not sure which one is preferred or if we need both (hence two PRs) 😀

What do you mean a tag? It should always be an error node.

magnus-madsen commented:
I am still confused about these two PRs and which should be merged when and whatnot.

What I expected was:

- Introduce RestrictablePattern.Error (or whatever it is called) and propagate it in every phase.
- Refactor the error to use Pattern.Error.

jaschdoc commented:
I understand.

There are two patterns: a `Var` pattern and a `Tag` pattern which are disjoint patterns. In both paths of the code the same error is used which is a problem since one branch returns a `HardFailure` and the other one returns a `SoftFailure` which places different type requirements on the error.

I decided to split the work into two distinct PRs that deal with each path. This one has the minimal amount of changes (the `Tag` path was a bit more complex). So once this is merged into main, we can merge the other PR with these changes which should arrive at what you wanted.

This is all just to introduce the changes incrementally and to avoid breaking things.

jaschdoc commented on main/src/ca/uwaterloo/flix/language/phase/Weeder2.scala:
```diff
@@ -1400,13 +1400,11 @@ object Weeder2 {
-          val patVar = WeededAst.RestrictableChoosePattern.Error(other.loc)
-          val qname = Name.mkQName(s"choosePattern${flix.genSym.freshId()}", other.loc.asSynthetic)
-          Validation.toSoftFailure(RestrictableChoosePattern.Tag(qname, List(patVar), other.loc.asSynthetic), e)
+          Validation.toSoftFailure(RestrictableChoosePattern.Error(other.loc), e)
```
this is the important change

jaschdoc commented:
RFR :)

magnus-madsen commented:
@mlutze Can you do a first review?

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/Lowering.scala:
```diff
@@ -968,6 +968,7 @@ object Lowering {
           val tagSym = visitRestrictableCaseSymUse(sym)
           val p = LoweredAst.Pattern.Tag(tagSym, pat1, tpe, loc)
           LoweredAst.MatchRule(p, None, e)
+        case TypedAst.RestrictableChoosePattern.Error(_, loc) => throw InternalCompilerException("unexpected error restrictable choose pattern near", loc)
```
Is it standard to use "near" in ICEs?

JonathanStarup replied:
I don't think so, especially since ICE doesn't print it location, so it will always just say `unexpected error restrictable choose pattern near`

magnus-madsen replied:
Not really. In some places we do it. We should probably think of something. 

The location is already there, so not super relevant to repeat.

JonathanStarup replied:
the location is not not printed though, so you cant see it

JonathanStarup replied:
throw ICE(msg, loc) and throw ICE(msg, SourceLocation.Unknown) gives the exact same information

jaschdoc replied:
So what do we agree on?

jaschdoc replied:
I'll remove `near` for now but we should figure something out. I'll bring this up next meeting, maybe just a quick vote?

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/RestrictableChooseConstraintGen.scala:
```diff
@@ -75,99 +76,115 @@ object RestrictableChooseConstraintGen {
-        val enumSym = rules0.headOption.getOrElse(throw InternalCompilerException("unexpected empty choose", loc)).pat match {
-          case KindedAst.RestrictableChoosePattern.Tag(sym, _, _, _) => sym.sym.enumSym
+        val enumSym0 = rules0.headOption.flatMap {
+          case rule => rule.pat match {
```
Can we document a bit what's going on here? This place is already confusing :)

magnus-madsen commented:
Lets fix Matts second comment and then we can go.

jaschdoc commented:
RFR

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/RestrictableChooseConstraintGen.scala:
```diff
@@ -75,99 +76,114 @@ object RestrictableChooseConstraintGen {
+          case None => throw InternalCompilerException("unexpected empty choose", loc)
+          case Some(rule) => rule.pat match {
+            case RestrictableChoosePattern.Error(_, loc) => // If the pattern is an error we just return fresh error variables
+              (Type.freshError(Kind.Error, loc), Type.freshError(Kind.Error, loc))
```
Do we use `Kind.Error` elsewhere like this? I thought `Kind.Error` is only when we don't know the kind(?)

jaschdoc replied:
I think I just assumed that this was the right thing to do. I can turn it into a `Star` instead :)

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/RestrictableChooseConstraintGen.scala:
```diff
@@ -75,99 +76,114 @@ object RestrictableChooseConstraintGen {
+            val resEff = Type.mkUnion(eff :: effs, loc)
+            (resTpe, resEff)
 
+          case None => (Type.freshError(Kind.Error, loc), Type.freshError(Kind.Error, loc))
```
Ditto

mlutze commented on main/src/ca/uwaterloo/flix/language/phase/typer/RestrictableChooseConstraintGen.scala:
```diff
@@ -75,99 +76,114 @@ object RestrictableChooseConstraintGen {
+            val resEff = Type.mkUnion(eff :: effs, loc)
+            (resTpe, resEff)
 
+          case None => (Type.freshError(Kind.Star, loc), Type.freshError(Kind.Star, loc))
```
The second one should actually be Kind.Eff, since that's the inferred effect.
